cpp/src/arrow/flight/client_cookie_middleware.cc

```diff
@@ -0,0 +1,327 @@
+      // Create month integer, pad with leading zeros if required.
+      std::string padded_month;
+      if ((i + 1) < 10) {
+        padded_month = "0";
+      }
+      padded_month += std::to_string(i + 1);
+
+      // Replace symbolic month with numeric month.
+      date->replace(it, months[i].length(), padded_month);
+
+      // String is now in format: '01 01 3000 22:15:36 GMT'.
+      break;
+    }
+  }
+
+  // String is now in format '01 01 3000 22:15:36'.
+  auto it = date->find(" GMT");
+  date->erase(it, 4);
+
+  // Sometimes a semicolon is added at the end, if this is the case, remove it.
+  if (date->back() == ';') {
+    date->pop_back();
+  }
+}
+
+struct Cookie {
+  static Cookie parse(const arrow::util::string_view& cookie_header_value) {
+    // Parse the cookie string. If the cookie has an expiration, record it.
+    // If the cookie has a max-age, calculate the current time + max_age and set that as
+    // the expiration.
+    Cookie cookie;
+    cookie.has_expiry_ = false;
+    std::string cookie_value_str(cookie_header_value);
+
+    // There should always be a first match which should be the name and value of the
+    // cookie.
+    std::string::size_type pos = 0;
+    CookiePair cookie_pair = ParseCookieAttribute(cookie_value_str, pos);
+    if (!cookie_pair.has_value()) {
+      // No cookie found. Mark the output cookie as expired.
+      cookie.has_expiry_ = true;
+      cookie.expiration_time_ = std::chrono::system_clock::now();
+    } else {
+      cookie.cookie_name_ = cookie_pair.value().first;
+      cookie.cookie_value_ = cookie_pair.value().second;
+    }
+
+    while (pos < cookie_value_str.size()) {
+      cookie_pair = ParseCookieAttribute(cookie_value_str, pos);
+      if (!cookie_pair.has_value()) {
+        break;
+      }
+
+      std::string cookie_attr_value_str = cookie_pair.value().second;
+      if (arrow::internal::AsciiEqualsCaseInsensitive(cookie_pair.value().first,
+                                                      "max-age")) {
+        // Note: max-age takes precedence over expires. We don't really care about other
+        // attributes and will arbitrarily take the first max-age. We can stop the loop
+        // here.
+        cookie.has_expiry_ = true;
+        int max_age = std::stoi(cookie_attr_value_str);
+        if (max_age <= 0) {
+          // Force expiration.
+          cookie.expiration_time_ = std::chrono::system_clock::now();
+        } else {
+          // Max-age is in seconds.
+          cookie.expiration_time_ =
+              std::chrono::system_clock::now() + std::chrono::seconds(max_age);
+        }
+        break;
+      } else if (arrow::internal::AsciiEqualsCaseInsensitive(cookie_pair.value().first,
+                                                             "expires")) {
+        cookie.has_expiry_ = true;
+        int64_t seconds = 0;
+        FixDate(&cookie_attr_value_str);
+        const char* COOKIE_EXPIRES_FORMAT = "%d %m %Y %H:%M:%S";
```

Comment:
Can this be moved to an actual constant following the naming convention (kCookieExpiresFormat or similar)?